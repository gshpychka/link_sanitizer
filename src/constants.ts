export const envVars = {
  urlBlackList: 'URL_PARAMETERS_BLACKLIST',
  messageQueueUrl: 'MESSAGE_QUEUE_URL',
  telegramTokenParameterName: 'TELEGRAM_TOKEN_PARAMETER_NAME',
};
