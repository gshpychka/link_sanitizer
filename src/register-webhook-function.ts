// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for RegisterWebhookFunction
 */
export interface RegisterWebhookFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/register-webhook.
 */
export class RegisterWebhookFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: RegisterWebhookFunctionProps) {
    super(scope, id, {
      description: 'src/register-webhook.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs16.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/register-webhook.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}