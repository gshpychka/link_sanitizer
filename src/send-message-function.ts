// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for SendMessageFunction
 */
export interface SendMessageFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/send-message.
 */
export class SendMessageFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: SendMessageFunctionProps) {
    super(scope, id, {
      description: 'src/send-message.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/send-message.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}