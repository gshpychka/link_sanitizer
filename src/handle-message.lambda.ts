import { parse as urlparse, UrlWithParsedQuery, format as urlformat } from 'url';
import { SQSClient, SendMessageCommand, SendMessageCommandInput } from '@aws-sdk/client-sqs';
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { Message, Update } from 'node-telegram-bot-api';
import { messageHandlerEnvVars } from './constants';

const sqs = new SQSClient({});

export const getUrlParamBlacklist = (): string[] => {
  return JSON.parse(process.env[messageHandlerEnvVars.urlBlackList]!);
};


export const sendCleanedUrls = async (originalMessage: Message, cleanedUrls: string[]) => {
  try {
    const messageQueueUrl = process.env[messageHandlerEnvVars.messageQueueUrl]!;
    let messageText: string;

    if (cleanedUrls.length === 1) {
      messageText = `Ось лінк без трекінгу: ${cleanedUrls}`;
    } else {
      messageText = `Ось лінки без трекінгу: ${cleanedUrls.join(', ')}`;
    }
    const message: SendMessageCommandInput = {
      QueueUrl: messageQueueUrl,
      MessageBody: JSON.stringify({
        chatId: originalMessage.chat.id,
        text: messageText,
        props: { reply_to_message_id: originalMessage.message_id },
      }),
    };
    const command = new SendMessageCommand(message);
    await sqs.send(command);
  } catch (err) {
    console.error('Error sending URLs');
    throw err;
  }
};

export const removeTrackingIfPresent = (url: string, parameterBlacklist: string[]): string | null => {
  try {

    let urlParsed: UrlWithParsedQuery = urlparse(url, true);

    let trackingPresent = false;
    const query_params = urlParsed.query;

    for (const blacklistedParameter of parameterBlacklist) {
      if (blacklistedParameter in query_params) {
        delete query_params[blacklistedParameter];
        trackingPresent = true;
      }
    }

    if (trackingPresent) {
      urlParsed.query = query_params;
      // need to set `search` to `null` because `url.format` uses it by default
      // instead of `query`
      urlParsed.search = null;
      return urlformat(urlParsed);
    } else {
      return null;
    }
  } catch (err) {
    console.error('Error removing tracking');
    throw err;
  }

};

export const extractUrls = (message: Message): string[] | null => {
  const urls = [];
  try {
    if (message?.entities) {
      for (const entity of message.entities) {
        if (entity.type == 'url') {

          urls.push(message.text!.slice(entity.offset, entity.offset + entity.length));
        }
      }
    }
  } catch (err) {
    console.error('Error extracting URLs');
  }
  return (urls.length > 0) ? urls : null;
};

export const processMessage = async (msg: Message, parameterBlacklist: string[]) => {
  const urls = extractUrls(msg);
  if (!urls) {
    console.log('Message does not contain a URL');
    return;
  }

  const cleanUrls = urls.map(
    url => removeTrackingIfPresent(
      url, parameterBlacklist,
    ),
  ).filter((x): x is string => x !== null);

  if (cleanUrls.length > 0) {
    console.log('Sending cleaned URLs');
    await sendCleanedUrls(msg, cleanUrls);
  } else {
    console.log('URLs do not need cleaning');
  }

};


export const handler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  try {

    console.log(event.body);

    const update: Update = JSON.parse(event.body!);
    const parameterBlacklist = getUrlParamBlacklist();

    if (update?.message) {
      await processMessage(update.message, parameterBlacklist);
    }

    return {
      isBase64Encoded: false,
      statusCode: 200,
      body: '0',
    };
  } catch (err) {
    console.error(`Error in handler: ${JSON.stringify(err)}`);
    return {
      isBase64Encoded: false,
      statusCode: 500,
      body: '0',
    };
  }

};
