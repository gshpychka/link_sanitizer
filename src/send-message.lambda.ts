import { SSMProvider } from '@aws-lambda-powertools/parameters/ssm';
import { SQSEvent } from 'aws-lambda';
import TelegramBot from 'node-telegram-bot-api';
import { envVars } from './constants';


const parametersProvider = new SSMProvider();

export const getTelegramToken = async (): Promise<string> => {
  try {
    const token = await parametersProvider.get(
      process.env[envVars.telegramTokenParameterName]!, { decrypt: true, maxAge: 60 * 60 * 24 },
    ) ?? null;
    if (!token) {
      throw new Error('Token is empty');
    } else {
      return token;
    }
  } catch (err) {
    console.error('Error getting parameter');
    throw err;
  }
};

export const handler = async (event: SQSEvent): Promise<void> => {
  const telegramToken = await getTelegramToken();
  const bot = new TelegramBot(telegramToken, { polling: false });
  for (const record of event.Records) {
    console.log(`Processing element: ${record.body}`);
    const message = JSON.parse(record.body);
    const sentMessage = await bot.sendMessage(
      message.chatId,
      message.text,
      message.props,
    );
    console.log(`Sent message ID: ${sentMessage.message_id}`);
  }
};
