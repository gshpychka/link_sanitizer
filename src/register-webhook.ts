import * as ssm from 'aws-cdk-lib/aws-ssm';
import * as cdk from 'aws-cdk-lib/core';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { RegisterWebhookFunction } from './register-webhook-function';

export interface TelegramTokenProps {
  parameter: ssm.IStringParameter;
  version?: string | undefined;
}

export interface TelegramBotWebhookProps {
  telegramToken: TelegramTokenProps;
  webhookUrl: string;
}


export interface TelegramWebhookResourceProperties {
  tokenParameterName: string;
  webhookURL: string;
}


export class TelegramBotWebhook extends Construct {
  constructor(scope: Construct, id: string, props: TelegramBotWebhookProps) {
    super(scope, id);

    const tokenParameterName = props.telegramToken.version ?
      `${props.telegramToken.parameter.parameterName}:${props.telegramToken.version}`
      : props.telegramToken.parameter.parameterName;

    const handler = new RegisterWebhookFunction(this, 'Handler');

    props.telegramToken.parameter.grantRead(handler);


    const provider = new cr.Provider(this, 'Provider', { onEventHandler: handler });

    new cdk.CustomResource(this, 'Default', {
      resourceType: 'Custom::TelegramBotWebhook',
      serviceToken: provider.serviceToken,
      properties: {
        tokenParameterName: tokenParameterName,
        webhookURL: props.webhookUrl,
      } satisfies TelegramWebhookResourceProperties,
    },
    );
  }

}
