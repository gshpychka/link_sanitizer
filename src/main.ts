import { App, Stack, StackProps } from 'aws-cdk-lib';
import * as apigw from 'aws-cdk-lib/aws-apigateway';
import * as event_sources from 'aws-cdk-lib/aws-lambda-event-sources';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';
import { envVars } from './constants';
import { HandleMessageFunction } from './handle-message-function';
import { SendMessageFunction } from './send-message-function';

interface LinkSanitizerProps extends StackProps {
  telegramTokenParameterName: string;
}
export class LinkSanitizer extends Stack {
  private readonly telegramToken: ssm.IStringParameter;
  constructor(scope: Construct, id: string, props: LinkSanitizerProps) {
    super(scope, id, props);

    this.telegramToken = ssm.StringParameter.fromSecureStringParameterAttributes(
      this, 'TelegramToken', {
        parameterName: props.telegramTokenParameterName,
      });

    const messageQueue = new sqs.Queue(this, 'MessageQueue');

    const messageHandler = new HandleMessageFunction(this, 'Handler', {
      environment: {
        [envVars.urlBlackList]: JSON.stringify(
          [
            'utm_source',
            'utm_medium',
            'utm_campaign',
            'utm_name',
            'utm_term',
            'utm_content',
            'gclid',
            'fbclid',
            'igshid',
            'mc_eid',
            'mc_cid',
            '_hsenc',
            '_hsmi',
            'hsCtaTracking',
            't',
          ]),
        [envVars.messageQueueUrl]: messageQueue.queueUrl,
      },
      retryAttempts: 0,
    });

    const messageSender = new SendMessageFunction(this, 'MessageSender', {
      retryAttempts: 0,
      environment: {
        [envVars.telegramTokenParameterName]: this.telegramToken.parameterName,
      },
    });

    this.telegramToken.grantRead(messageSender);

    messageQueue.grantSendMessages(messageHandler);

    const queueSource = new event_sources.SqsEventSource(messageQueue,
      { batchSize: 1, maxConcurrency: 2 },
    );
    messageSender.addEventSource(queueSource);
    new apigw.LambdaRestApi(this, 'Api', { handler: messageHandler, proxy: true });
  }
}

const app = new App();

new LinkSanitizer(app, 'LinkSanitizer', {
  telegramTokenParameterName: '/link-sanitizer/telegram-token',
});

app.synth();
