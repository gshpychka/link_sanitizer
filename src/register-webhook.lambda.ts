import crypto from 'crypto';
import { SSMProvider } from '@aws-lambda-powertools/parameters/ssm';
import { CdkCustomResourceResponse } from 'aws-lambda';
import TelegramBot from 'node-telegram-bot-api';
import { TelegramWebhookResourceProperties } from './register-webhook';


const parametersProvider = new SSMProvider();

export const getTelegramToken = async (parameterName: string): Promise<string> => {
  try {
    const token = await parametersProvider.get(
      parameterName, { decrypt: true, maxAge: 60 * 60 * 24 },
    ) ?? null;
    if (!token) {
      throw new Error('Token is empty');
    } else {
      return token;
    }
  } catch (err) {
    console.error('Error getting parameter');
    throw err;
  }
};

interface CustomResourceEvent {
  ResourceProperties: TelegramWebhookResourceProperties;
  RequestType: 'Create' | 'Update' | 'Delete';
}


export async function handler(event: CustomResourceEvent): Promise<CdkCustomResourceResponse> {
  console.log('Received event:', JSON.stringify(event, null, 2));

  const properties: TelegramWebhookResourceProperties = event.ResourceProperties;

  try {
    const token = await getTelegramToken(properties.tokenParameterName);
    const physicalId = crypto.createHash('sha256').update(token).digest('hex');

    const bot = new TelegramBot(token);

    let response: any;

    if (event.RequestType === 'Delete') {
      response = await bot.deleteWebHook();
    } else {
      response = await bot.setWebHook(properties.webhookURL);
    }

    if (!response) {
      throw new Error(`Failed to ${event.RequestType.toLowerCase()} webhook`);
    }

    return {
      PhysicalResourceId: physicalId,
    };
  } catch (error) {
    console.error('Error:', error);
    throw error;
  }
}
