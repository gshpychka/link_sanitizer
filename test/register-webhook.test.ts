import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { TelegramBotWebhook } from '../src/register-webhook';

test('TelegramBotWebhook', () => {
  const app = new App();
  const tokenParameterName = 'MyToken';
  const stack = new Stack(app, 'MyStack');

  const handlerFnId = 'WebhookHandler0697217E';
  const providerFnId = 'WebhookProviderframeworkonEventA9527C6B';
  const webhookUrl = 'example.com';

  const token = ssm.StringParameter.fromSecureStringParameterAttributes(stack, 'MyToken', { parameterName: tokenParameterName });

  new TelegramBotWebhook(stack, 'Webhook', { webhookUrl: webhookUrl, telegramToken: { parameter: token } });

  const template = Template.fromStack(stack);


  // Check the snapshot
  expect(template.toJSON()).toMatchSnapshot();

  // Check there are two Lambda functions created
  // Webhook Creator Provider & Handler
  const lambdas = template.findResources('AWS::Lambda::Function', {});
  expect(Object.keys(lambdas)).toHaveLength(2);

  const handlerLambda = lambdas[handlerFnId];
  expect(handlerLambda).toBeDefined();

  const providerLambda = lambdas[providerFnId];
  expect(providerLambda).toBeDefined();

  template.hasResource('Custom::TelegramBotWebhook', {
    Properties: {
      ServiceToken: { 'Fn::GetAtt': [providerFnId, 'Arn'] },
      tokenParameterName: tokenParameterName,
      webhookURL: webhookUrl,
    },
  });

  const handlerRolePolicies = template.findResources('AWS::IAM::Policy', {
    Properties: {
      Roles: [
        {
          Ref: handlerLambda.Properties.Role['Fn::GetAtt'][0],
        },
      ],
    },
  });

  interface PolicyStatement {
    Action: string | string[];
    Resource: any;
  }


  const handlerHasReadAccessToToken = Object.values(handlerRolePolicies).some(
    policy => {
      return policy.Properties.PolicyDocument.Statement.some((statement: PolicyStatement) => {
        const actions = Array.isArray(statement.Action) ? statement.Action : [statement.Action];
        return actions.includes('ssm:GetParameter') &&
          JSON.stringify(statement.Resource) === JSON.stringify(
            {
              'Fn::Join': [
                '',
                [
                  'arn:',
                  {
                    Ref: 'AWS::Partition',
                  },
                  ':ssm:',
                  {
                    Ref: 'AWS::Region',
                  },
                  ':',
                  {
                    Ref: 'AWS::AccountId',
                  },
                  `:parameter/${tokenParameterName}`,
                ],
              ],
            });
      },
      );
    });

  expect(handlerHasReadAccessToToken).toBe(true);

});
