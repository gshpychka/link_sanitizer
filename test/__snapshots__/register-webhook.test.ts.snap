// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TelegramBotWebhook 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Webhook": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "WebhookProviderframeworkonEventA9527C6B",
            "Arn",
          ],
        },
        "tokenParameterName": "MyToken",
        "webhookURL": "example.com",
      },
      "Type": "Custom::TelegramBotWebhook",
      "UpdateReplacePolicy": "Delete",
    },
    "WebhookHandler0697217E": {
      "DependsOn": [
        "WebhookHandlerServiceRoleDefaultPolicy5B9DDF0F",
        "WebhookHandlerServiceRole82D8E58D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "be72627956082b34daea4dc38d5366dc878f5f239ab70fcac30e739998fa625f.zip",
        },
        "Description": "src/register-webhook.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "WebhookHandlerServiceRole82D8E58D",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebhookHandlerServiceRole82D8E58D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebhookHandlerServiceRoleDefaultPolicy5B9DDF0F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/MyToken",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebhookHandlerServiceRoleDefaultPolicy5B9DDF0F",
        "Roles": [
          {
            "Ref": "WebhookHandlerServiceRole82D8E58D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WebhookProviderframeworkonEventA9527C6B": {
      "DependsOn": [
        "WebhookProviderframeworkonEventServiceRoleDefaultPolicy4B4F71B4",
        "WebhookProviderframeworkonEventServiceRole483AC2AF",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f2d30cfc360482320a52a4fcde8a70f3569df79ab30be24650fda58eb60052cf.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (MyStack/Webhook/Provider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "WebhookHandler0697217E",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "WebhookProviderframeworkonEventServiceRole483AC2AF",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebhookProviderframeworkonEventServiceRole483AC2AF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebhookProviderframeworkonEventServiceRoleDefaultPolicy4B4F71B4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "WebhookHandler0697217E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "WebhookHandler0697217E",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WebhookProviderframeworkonEventServiceRoleDefaultPolicy4B4F71B4",
        "Roles": [
          {
            "Ref": "WebhookProviderframeworkonEventServiceRole483AC2AF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
